### PR TITLE
Infer types in list() over tuple

### PR DESCRIPTION
**Pull request by i582**
Added type inference for each variable when destructuring a variable with the `\tuple` type or a function that returns `\tuple`.
Added test.

For example:
```php
class Boo {}

/**
 * @return \tuple(int, \Boo, int, string)
 */
function foo() {
    return [5, new Boo(), 10, "gas"];
}

$tuple = foo();

[$i, $bo, $j, $str] = $tuple;

exprtype($i, "int");
exprtype($bo, "\Boo");
exprtype($j, "int");
exprtype($str, "string");
```

### CODE REVIEW

File: src/linter/block.go [hunk 1]
```diff
@@ -2313,7 +2313,7 @@ func (b *BlockWalker) handleAssign(a *assign.Assign) bool {
 			return true
 		}
 
-		tp := solver.ExprType(b.ctx.sc, b.r.ctx.st, a.Expression)
+		tp := solver.ExprTypeLocal(b.ctx.sc, b.r.ctx.st, a.Expression)
```
Comment (i582) [inline src/linter/block.go:2316]: but there is no race, ahaha

File: src/linter/block.go [hunk 2]
```diff
@@ -2302,7 +2309,19 @@ func (b *BlockWalker) handleAssign(a *assign.Assign) bool {
...
 	case *expr.List:
-		b.handleAssignList(v.Items)
+		if !meta.IsIndexingComplete() {
+			return true
+		}
+
+		tp := solver.ExprTypeLocal(b.ctx.sc, b.r.ctx.st, a.Expression)
+		var findType string
+		tp.Iterate(func(t string) {
+			if strings.HasPrefix(t, "\\shape") {
```
Comment (quasilyte) [inline src/linter/block.go:2319]: You can use raw string literal to avoid `\` escaping:
```
`\shape`
```

File: src/linter/block.go [hunk 3]
```diff
@@ -2302,7 +2309,19 @@ func (b *BlockWalker) handleAssign(a *assign.Assign) bool {
 		b.checkVoidType(a.Expression)
 		b.replaceVar(v, solver.ExprTypeLocal(b.ctx.sc, b.r.ctx.st, a.Expression), "assign", meta.VarAlwaysDefined)
 	case *expr.List:
-		b.handleAssignList(v.Items)
+		if !meta.IsIndexingComplete() {
+			return true
+		}
+
+		tp := solver.ExprTypeLocal(b.ctx.sc, b.r.ctx.st, a.Expression)
+		var findType string
```
Comment (quasilyte) [inline src/linter/block.go:2317]: `findType` is a verb. It's not a good variable name here in my opinion.
You can call it `shapeType` given that it's only assigned to shape types.

File: src/linter/block.go [hunk 4]
```diff
@@ -2302,7 +2309,19 @@ func (b *BlockWalker) handleAssign(a *assign.Assign) bool {
...
+		tp.Iterate(func(t string) {
+			if strings.HasPrefix(t, "\\shape") {
+				findType = t
+			}
+		})
+		class, _ := meta.Info.GetClass(findType)
```
Comment (quasilyte) [inline src/linter/block.go:2323]: If findType (or shapeType) is an empty string, there is no need to do a lookup.

File: src/linter/block.go [hunk 5]
```diff
@@ -2196,9 +2196,16 @@ func (b *BlockWalker) handleAssignReference(a *assign.Reference) bool {
...
+func (b *BlockWalker) handleAssignList(items []*expr.ArrayItem, info meta.ClassInfo) {
+	for i, item := range items {
+		prop, ok := info.Properties[fmt.Sprint(i)]
+		var tp meta.TypesMap
+		if !ok {
+			tp = meta.NewTypesMap("unknown_from_list")
+		} else {
+			tp = prop.Typ
+		}
+		b.handleVariableNode(item.Val, tp, "assign")
```
Comment (quasilyte) [inline src/linter/block.go:2208]: This method will handle all types like shape types. For non-shape types it's easier to keep the previous loop and avoid doing `fmt.Sprint` and map lookup per every item.

**Comment by i582** [comment]
closed due #602

**Comment by quasilyte** [comment]
Race condition is after racing you. :sweat_smile:

**Pull request closed without merging**